File: java/xrayCalculations/Source.java
package fpSimulation;

public class Source {
    double keVsrc;
    double milliamps;
    double steradians;
    double thetaIn;
    double thetaOut;
    double contMult;
    double charMult;
    
    int z = 45;
    double bremC = 1.36 * Math.pow(10,9);
    double bremX;
    double angleTerm;
    double ATerm = 102.906;
    double mTerm;
    double JTerm;
    double nTerm;
    double rhozmTerm;
    double rhozbarTerm;
    double[][] I;
    
    public Source(double keVsrc, double milliamps, double steradians, double thetaIn, double thetaOut, double contMult, double charMult) {
        this.keVsrc = keVsrc;
        this.milliamps = milliamps;
        this.steradians = steradians;
        this.thetaIn = thetaIn;
        this.thetaOut = thetaOut;
        this.contMult = contMult;
        this.charMult = charMult;
        
        this.bremX = 1.0314 - 0.0032*this.z + 0.0047*this.keVsrc;
        this.angleTerm = Math.sin(Math.toRadians(this.thetaIn))/Math.sin(Math.toRadians(this.thetaOut));
        this.mTerm = 0.1382 - (0.9211/Math.pow(this.z,0.5));
        this.JTerm = 0.0135 * this.z;
        this.nTerm = Math.pow(this.keVsrc,this.mTerm) * ( 0.1904 - 0.2236*Math.log(this.z) + 0.1292*Math.pow(Math.log(this.z),2) - 0.0149*Math.pow(Math.log(this.z),3) );
        this.rhozmTerm = (this.ATerm/this.z) * ( 0.787*Math.pow(10,-5)*Math.sqrt(this.JTerm)*Math.pow(this.keVsrc,1.5) + 0.735*Math.pow(10,-6)*Math.pow(this.keVsrc,2) );
        this.rhozbarTerm = (0.49269 - 1.0987*this.nTerm + 0.78557*Math.pow(this.nTerm,2));
        
        Energies energies = new Energies(1.00,50.00,0.02);
        int nEnergies = energies.keV.length;
        this.I = new double[nEnergies][2];
        for (int n = 0; n < nEnergies; n++) {
            this.I[n][0] = energies.keV[n];
            this.I[n][1] = 0.0;
        }
    }
    
    public double[][] calculateIntensity(double[][] tau, double[][] RhPeaks) {
        double totalI;
        double contI;
        double charI;
        double keVi;
        double tauAtEi;
        
        double shell;
        double keVline;
        double keVedge;
        double fYield;
        double pTrans;
        
        int nEnergies = this.I.length;
        int nPeaks = RhPeaks.length;
        for (int n = 0; n < nEnergies; n++) {
            keVi = this.I[n][0];
            tauAtEi = tau[n][this.z+1];
            if (keVi <= this.keVsrc) {
                contI = calculateContEi(keVi, tauAtEi);
                totalI = contI;
                for (int p = 0; p < nPeaks; p++) {
                    shell = RhPeaks[p][0];
                    keVline = RhPeaks[p][1];
                    keVedge = RhPeaks[p][2];
                    fYield = RhPeaks[p][3];
                    pTrans = RhPeaks[p][4];
                    if ( (keVedge <= this.keVsrc) && matchEi(keVi, keVline) ) {
                        charI = calculateCharEi(keVi, tauAtEi, shell, keVedge, fYield, pTrans);
                        totalI = totalI + charI;
                    }
                }
                if (totalI > 1) {
                    this.I[n][1] = totalI;
                }
            } else {
                break;
            }
        }
        return this.I;
    }
    
    private double calculateContEi(double keVi, double tauAtEi) {
        double Uo = this.keVsrc / keVi;
        double lnUo = Math.log(Uo);
        double rhozbar = (this.rhozmTerm * this.rhozbarTerm * lnUo) / (0.70256 - 1.09865*this.nTerm + 1.0046*Math.pow(this.nTerm,2) + lnUo);
        double corrTerm = 2 * tauAtEi * rhozbar * this.angleTerm;
        double fChi = (1 - Math.exp(-corrTerm)) / corrTerm;
        fChi = 1 - (1-fChi)*this.contMult;
        
        double bremI = 0.02 * (this.steradians * this.milliamps * this.bremC * this.z * Math.pow(Uo-1,this.bremX));
        double contI = bremI * fChi;
        return contI;
    }
    
    private double calculateCharEi(double keVi, double tauAtEi, double shell, double keVedge, double fYield, double pTrans) {
        double Uo = this.keVsrc / keVedge;
        double lnUo = Math.log(Uo);
        double rhozbar = (this.rhozmTerm * this.rhozbarTerm * lnUo) / (0.70256 - 1.09865*this.nTerm + 1.0046*Math.pow(this.nTerm,2) + lnUo);
        double corrTerm = 2 * tauAtEi * rhozbar * this.angleTerm;
        double fChi = (1 - Math.exp(-corrTerm)) / corrTerm;
        fChi = 1 - (1-fChi)*this.contMult;
        
        double invStopPow = 1 + 16.05 * Math.sqrt(this.JTerm/keVedge) * ( ( Math.sqrt(Uo)*lnUo + 2*(1-Math.sqrt(Uo)) ) / (Uo*lnUo + 1 - Uo) );
        double invStopRow = (1/this.z) * (Uo*lnUo + 1 - Uo) * invStopPow;
        double backScatter = 1 - 0.0081517*this.z + 3.613*Math.pow(10,-5)*Math.pow(this.z,2) + 0.009583*this.z*Math.exp(-Uo) + 0.001141*this.keVsrc;
        
        double charI = this.steradians * this.milliamps * invStopPow * backScatter * fYield * pTrans * fChi;
        if (shell == 1.0) {
            charI = (2.0 * 0.35 * 5.0*Math.pow(10,13)) * charI;
        } else {
            charI = (8.0 * 0.25 * 6.9*Math.pow(10,13)) * charI;
        }
        charI = charI * this.charMult;
        return charI;
    }
    
    private boolean matchEi(double keVi, double keVpeak) {
        boolean match = false;
        int eVi = (int) (keVi*100);
        int eVpeak = (int) (keVpeak*100);
        if ((eVpeak % 2) != 0) {
            eVpeak = eVpeak + 1;
        }
        if (eVi == eVpeak) {
            match = true;
        }
        return match;
    }
}

File: go/src/fpSimulation/source.go
package fpSimulation

import (
	"math"
)

func sourceIntensity(scen *Scenario, prop *Properties) {
	var contI, charI float64
	var keVi, keVline, keVedge float64
	var nEnergies, nPeaks int

	co := sourceConstants(scen)

	nEnergies = len(scen.SrcI)
	for k := 0; k < nEnergies; k++ {
		keVi = scen.SrcI[k][0]
		if keVi <= scen.KeVsrc {
			contI = calculateContEi(k, co, scen, prop)
			if contI > 1 {
				scen.SrcI[k][1] = contI
			}
		} else {
			break
		}
	}
	nPeaks = len(prop.Rh)
	for p := 0; p < nPeaks; p++ {
		keVline = prop.Rh[p][1]
		keVedge = prop.Rh[p][2]
		if keVedge <= scen.KeVsrc {
			keVi := closestEi(keVline)
			if keVi <= 50.0 {
				k := rowFromE(keVi)
				charI = calculateCharEi(k, p, co, scen, prop)
				scen.SrcI[k][1] = scen.SrcI[k][1] + charI
			}
		}
	}
	return
}

func sourceConstants(scen *Scenario) *SourceConst {
	co := new(SourceConst)
	co.zInt = 45
	co.zFlt = 45.0
	co.bremC = 1.36 * math.Pow(10.0, 9.0)
	co.bremX = 1.0314 - 0.0032*co.zFlt + 0.0047*scen.KeVsrc
	co.angleTerm = math.Sin(scen.ThetaIn*(math.Pi/180.0)) / math.Sin(scen.ThetaOut*(math.Pi/180.0))
	co.ATerm = 102.906
	co.mTerm = 0.1382 - (0.9211 / math.Pow(co.zFlt, 0.5))
	co.JTerm = 0.0135 * co.zFlt
	co.nTerm = math.Pow(scen.KeVsrc, co.mTerm) * (0.1904 - 0.2236*math.Log(co.zFlt) + 0.1292*math.Pow(math.Log(co.zFlt), 2.0) - 0.0149*math.Pow(math.Log(co.zFlt), 3.0))
	co.rhozmTerm = (co.ATerm / co.zFlt) * (0.787*math.Pow(10.0, -5.0)*math.Sqrt(co.JTerm)*math.Pow(scen.KeVsrc, 1.5) + 0.735*math.Pow(10.0, -6.0)*math.Pow(scen.KeVsrc, 2.0))
	co.rhozbarTerm = (0.49269 - 1.0987*co.nTerm + 0.78557*math.Pow(co.nTerm, 2.0))
	return co
}

func calculateContEi(k int, co *SourceConst, scen *Scenario, prop *Properties) float64 {
	var keVi, tauAtEi, Uo, lnUo, rhozbar, corrTerm, fChi, bremI, contI float64

	keVi = scen.SrcI[k][0]
	tauAtEi = prop.Tau[k][co.zInt+1]

	Uo = scen.KeVsrc / keVi
	lnUo = math.Log(Uo)
	rhozbar = (co.rhozmTerm * co.rhozbarTerm * lnUo) / (0.70256 - 1.09865*co.nTerm + 1.0046*math.Pow(co.nTerm, 2.0) + lnUo)
	corrTerm = 2.0 * tauAtEi * rhozbar * co.angleTerm
	fChi = (1.0 - math.Exp(-corrTerm)) / corrTerm
	fChi = 1.0 - (1.0-fChi)*scen.ContMult

	bremI = 0.02 * (scen.Sr * scen.MA * co.bremC * co.zFlt * math.Pow(Uo-1.0, co.bremX))
	contI = bremI * fChi
	return contI
}

func calculateCharEi(k int, p int, co *SourceConst, scen *Scenario, prop *Properties) float64 {
	var tauAtEi, shell, keVedge, fYield, pTrans, Uo, lnUo, rhozbar, corrTerm, fChi float64
	var invStopPow, backScatter, charI float64

	tauAtEi = prop.Tau[k][co.zInt+1]
	shell = prop.Rh[p][0]
	keVedge = prop.Rh[p][2]
	fYield = prop.Rh[p][3]
	pTrans = prop.Rh[p][4]

	Uo = scen.KeVsrc / keVedge
	lnUo = math.Log(Uo)
	rhozbar = (co.rhozmTerm * co.rhozbarTerm * lnUo) / (0.70256 - 1.09865*co.nTerm + 1.0046*math.Pow(co.nTerm, 2.0) + lnUo)
	corrTerm = 2.0 * tauAtEi * rhozbar * co.angleTerm
	fChi = (1.0 - math.Exp(-corrTerm)) / corrTerm
	fChi = 1.0 - (1.0-fChi)*scen.ContMult

	invStopPow = 1.0 + 16.05*math.Sqrt(co.JTerm/keVedge)*((math.Sqrt(Uo)*lnUo+2.0*(1.0-math.Sqrt(Uo)))/(Uo*lnUo+1.0-Uo))
	invStopPow = (1.0 / co.zFlt) * (Uo*lnUo + 1.0 - Uo) * invStopPow
	backScatter = 1 - 0.0081517*co.zFlt + 3.613*math.Pow(10.0, -5.0)*math.Pow(co.zFlt, 2.0) + 0.009583*co.zFlt*math.Exp(-Uo) + 0.001141*scen.KeVsrc

	charI = scen.Sr * scen.MA * invStopPow * backScatter * fYield * pTrans * fChi
	if shell == 1.0 {
		charI = (2.0 * 0.35 * 5.0 * math.Pow(10.0, 13.0)) * charI
	} else {
		charI = (8.0 * 0.25 * 6.9 * math.Pow(10.0, 13.0)) * charI
	}
	charI = charI * scen.CharMult
	return charI
}

func matchEi(keVi float64, keVpeak float64) bool {
	var match bool
	var eVi, eVpeak int

	match = false
	eVi = (int)(keVi * 100)
	eVpeak = (int)(keVpeak * 100)
	if (eVpeak % 2) != 0 {
		eVpeak = eVpeak + 1
	}
	if eVi == eVpeak {
		match = true
	}
	return match
}
